File: container_templates/matlab/container_example/src/algorithm_example.m
% algorithm_example.md - template to read/write to local directory 
% or S3 bucket
% ---------------------------------------------------------------------
%     
% Example:
%  
%  in terminal -
%  matlab -nodisplay -r "cd src; algorithm_example <manifest_in path>"
%  
%  in MATLAB console - 
%  setenv("PROCESSING_DROPBOX","<dropbox path>");
%  algorithm_example(<manifest_in path>)
% 
% Notes:
%   - If running from the terminal add
%   export PROCESSING_DROPBOX="<dropbox path>" to your bash profile
%   - <path> may either be the path to the local directory e.g.
%   '../container_example_data/dropbox/input_manifest_20220923t000000_local.json' 
%   or the path to the s3 bucket (e.g. s3://bucketname/).
%   - For instructions on using this code with Docker refer to
%   docker.md
%
% Authors: Laura Sandoval, Matt Watwood, Gavin Medley
%
% External libraries used:
%
% Stefan Stoll (2022). MD5 signature of a file 
% (https://www.mathworks.com/matlabcentral/fileexchange/5498-md5-signature-of-a-file), 
% MATLAB Central File Exchange. Retrieved November 9, 2022.
% 
% Matthew Spaethe (2022). Matlab logging facility 
% (https://www.mathworks.com/matlabcentral/fileexchange/42078-matlab-logging-facility), 
% MATLAB Central File Exchange. Retrieved November 15, 2022.
% ---------------------------------------------------------------------

function []=algorithm_example(manifest_in, varargin)
% Requires 1 input and has optional input.

addpath('logging')
import logging.*

% get the global LogManager
LogManager.getLogManager();

% add a logger instance to this script
logger = Logger.getLogger('AlgorithmExample');
logger.setLevel( Level.ALL );

defaultInt = 2;

logger.info(strcat('Command executed in: ', mfilename('filename'),'.m'));
logger.info('Parsing CLI arguments.');

p = inputParser;
addRequired(p,'manifest_in',@ischar);
addOptional(p,'multiplier',defaultInt,@isnumeric);
parse(p,manifest_in,varargin{:});

logger.info(strcat('Manifest file to read: ', p.Results.manifest_in));
logger.info(strcat('Additional options passed are multiplier = ', ...
    num2str(p.Results.multiplier)));

% only want 1 optional input at most
numvarargs = length(varargin);
if numvarargs > 1
    error('my_script:TooManyInputs', ...
        'requires at most 1 optional inputs');
end

processing_dropbox = getenv('PROCESSING_DROPBOX');

str = fileread(p.Results.manifest_in);
json_content = jsondecode(str);

logger.info(strcat('Manifest type is ', json_content.manifest_type));
logger.info(strcat('Manifest contains ', json_content.files.filename));

% Get the data from each file and put data into some format that will be 
% used in the algorithm
for n = 1 : length(json_content.files)

    checksum = json_content.files(n).checksum;
    filename = json_content.files(n).filename;

    assert(isequal(checksum,md5(filename)), ...
        'algorithm_example:checksumError','Checksums do not match!')
    logger.info(strcat('Checksum matches for ', filename));

  data_in = h5read(filename,'/HDFEOS/SWATHS/Swath1/DataField/Temperature');
  logger.info(strcat('Found input data in HDF5 file:',mat2str(data_in)))

end

%fake data product to write as output
data_out = data_in * p.Results.multiplier;

%Create hdf5 data
output_filepath = append(processing_dropbox, 'example_output.h5');
h5create(output_filepath,'/data/array1',[height(data_out) width(data_out)]);

logger.info(strcat('Writing output file: ', output_filepath))

h5write(output_filepath, '/data/array1', data_out);
h5writeatt(output_filepath,'/','someattr','hello, world');

% Need this to create empty group; if you are creating a group 
% that contains data, then use h5create and h5write
plist = 'H5P_DEFAULT';
fid = H5F.open(output_filepath,'H5F_ACC_RDWR',plist);
gid = H5G.create(fid,'new_group',plist,plist,plist);
H5G.close(gid);
H5F.close(fid);

json_out = append(processing_dropbox, 'output_manifest_20220923t111111.json');

logger.info(strcat('Writing output manifest: ', json_out))

%Get checksum of written file
checksum_written = md5(output_filepath);

s1.filename = output_filepath;
s1.checksum = checksum_written;
struct_1 = jsonencode(s1);

s.manifest_type = "OUTPUT";
s.files = [struct_1];
s.configuration = string(nan);

fid = fopen(json_out,'w');

fprintf(fid,'%s',strrep(jsonencode(s, "PrettyPrint", true),'\',''));
fclose(fid);

logger.info('Algorithm complete. Exiting.')

exit
end

File: container_templates/python/container_example/src/algorithm_example.py
"""
Container example for Level 2 developers

Authors: Laura Sandoval, Gavin Medley, Matt Watwood
"""
# Standard
import argparse
from hashlib import md5
import logging
import os
import sys
# Installed
import h5py as h5
import numpy as np
import pandas as pd
# Local
from libera_utils.io.smart_open import smart_open
from libera_utils.io.manifest import Manifest, ManifestType
from static.l2_quality_flags import L2QualityFlag


logger = logging.getLogger(__name__)


def main():
    # Initialize logging. You can configure logging however you like
    logging.basicConfig(level="DEBUG")

    # Get logger instance, named for the current function for traceability
    logger.debug(f"Command executed in container: {sys.argv}")
    logger.info("Parsing CLI arguments.")
    args = parse_cli_args(sys.argv[1:])

    manifest = args.manifest
    logger.info(f"Manifest file to read: {manifest}")

    logger.info("Additional options passed are "
                f"exampleint={args.exampleint}, examplefloat={args.examplefloat}, examplebool={args.examplebool}")

    processing_dropbox = os.environ['PROCESSING_DROPBOX']

    # read json information
    manifest = Manifest.from_file(manifest)
    if not manifest.manifest_type == ManifestType.INPUT:
        raise ValueError("Incorrect manifest type received as input.")

    logger.debug(f"Manifest type is {manifest.manifest_type}")
    logger.debug(f"Manifest contains files:\n{manifest.files}")

    # read hdf5
    for record in manifest.files:
        checksum = record['checksum']
        filename = record['filename']
        # Validate checksums
        with smart_open(filename, 'rb') as fh:
            if checksum != md5(fh.read()).hexdigest():
                raise ValueError("Checksums do not match!")
            logger.debug(f"Checksum matches for {filename}")

        with h5.File(smart_open(filename, 'rb'), 'r') as h5file:
            # Get the data from each file and put data into some format that will be used in the algorithm
            data_in = np.array(h5file['HDFEOS/SWATHS/Swath1/DataField/Temperature'])
            logger.info(f"Found input data in HDF5 file:\n{data_in}")

    df = generate_example_data()

    # write example output data to a new HDF5 file
    output_files = []
    output_filepath = os.path.join(processing_dropbox, 'example_output.h5')
    logger.info(f"Writing output file: {output_filepath}")
    with h5.File(smart_open(output_filepath, 'xb'), 'x') as hdf:
        hdf.create_group('new_group')
        hdf.attrs['someattr'] = "hello, world"
        hdf.create_dataset('data/array1', data=df.data_out)
        hdf.create_dataset('quality_out/array1', data=df.quality_int)

    # get the checksum of the written file
    with smart_open(output_filepath, 'rb') as fh:
        checksum = md5(fh.read()).hexdigest()

    output_files.append({"filename": output_filepath, "checksum": checksum})

    # Write output manifest file containing a list of the product files that the processing created
    output_manifest_path = os.path.join(processing_dropbox, "libera_output_manifest_20220923t111111.json")
    logger.info(f"Writing output manifest: {output_filepath}")
    output_manifest = Manifest(manifest_type=ManifestType.OUTPUT,
                               filename=output_manifest_path,
                               files=output_files,
                               configuration={})
    output_manifest.write(output_manifest_path)
    logger.info("Algorithm complete. Exiting.")


def generate_example_data():
    """
    Function created to make up data and quality flags

    Returns
    -------
    pandas.core.frame.DataFrame
        Dataframe containing data and quality flag columns
    """

    # generate fake data product to write as output
    df = pd.DataFrame()
    df['data_out'] = [-np.inf, -np.inf, 30, 40, 50, 60]
    df['quality'] = np.full(len(df.data_out), fill_value=L2QualityFlag.NONE)
    df['quality_int'] = np.full(len(df.data_out), fill_value=0)

    # flag neg infinite values
    was_inf = np.equal(df.data_out, -np.inf)
    df.loc[was_inf, 'quality'] = df.loc[was_inf, 'quality'][0] | L2QualityFlag.INF | L2QualityFlag.NEG
    df.loc[was_inf, 'quality_int'] = (df.loc[was_inf, 'quality'][0] | L2QualityFlag.INF | L2QualityFlag.NEG).value

    # view decomposed flag and value
    view_summary = df.quality.values[0].summary
    view_decomposed = (df.quality.values[0]).decompose()
    print(view_summary)
    print(view_decomposed)

    return df


def parse_cli_args(cli_args: list):
    """
    Function that parses CLI arguments

    Parameters
    ----------
    cli_args : list
        List of string arguments to parse

    Returns
    -------
    Namespace
        A Namespace object containing the parsed arguments as attributes
    """
    parser = argparse.ArgumentParser(description='program arguments.')

    parser.add_argument('--exampleint',
                        type=int,
                        default=42,
                        help='An example integer input option'
                        )

    parser.add_argument('--examplefloat',
                        type=float,
                        default=3.14159,
                        help='An example float input option')

    parser.add_argument('--examplebool',
                        type=bool,
                        default=True,
                        help='An example boolean input option')

    parser.add_argument('manifest',
                        type=str,
                        help="Input JSON manifest file containing a list of files to use in the processing algorithm.")

    return parser.parse_args(cli_args)


if __name__ == "__main__":
    main()